compiler/rustc_hir_typeck/src/demand.rs

```diff
@@ -795,6 +798,108 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+    /// parameter, instead of mutating the local binding. When encountering this we suggest:
+    ///
+    /// ```text
+    /// fn change_object(a: &'_ mut Ty) {
+    ///     let a = Ty::new();
+    ///     *b = a;
+    /// }
+    /// ```
+    fn annotate_mut_binding_to_immutable_binding(
+        &self,
+        err: &mut Diag<'_>,
+        expr: &hir::Expr<'_>,
+        error: Option<TypeError<'tcx>>,
+    ) -> bool {
+        let Some(TypeError::Sorts(ExpectedFound { expected, found })) = error else { return false };
+        let ty::Ref(_, inner, hir::Mutability::Not) = expected.kind() else { return false };
+        if !self.can_eq(self.param_env, *inner, found) {
+            // The difference between the expected and found values isn't one level of borrowing.
+            return false;
+        }
+        // We have an `ident = expr;` assignment.
+        let hir::Node::Expr(hir::Expr { kind: hir::ExprKind::Assign(lhs, rhs, _), .. }) =
+            self.tcx.parent_hir_node(expr.hir_id)
+        else {
+            return false;
+        };
+        if rhs.hir_id != expr.hir_id || expected.is_closure() {
+            return false;
+        }
+        // We are assigning to some binding.
+        let hir::ExprKind::Path(hir::QPath::Resolved(
+            None,
+            hir::Path { res: hir::def::Res::Local(hir_id), .. },
+        )) = lhs.kind
+        else {
+            return false;
+        };
+        let hir::Node::Pat(pat) = self.tcx.hir_node(*hir_id) else { return false };
+        // The pattern we have is an fn argument.
+        let hir::Node::Param(hir::Param { ty_span, .. }) = self.tcx.parent_hir_node(pat.hir_id)
+        else {
+            return false;
+        };
+        let item = self.tcx.hir().get_parent_item(pat.hir_id);
+        let item = self.tcx.hir_owner_node(item);
+        let Some(fn_decl) = item.fn_decl() else { return false };
+
+        // We have a mutable binding in the argument.
+        let hir::PatKind::Binding(hir::BindingMode::MUT, _hir_id, ident, _) = pat.kind else {
+            return false;
+        };
+
+        // Look for the type corresponding to the argument pattern we have in the argument list.
+        let Some(ty_sugg) = fn_decl
+            .inputs
+            .iter()
+            .filter_map(|ty| {
+                if ty.span == *ty_span
+                    && let hir::TyKind::Ref(lt, mut_ty) = ty.kind
+                {
+                    // `&'name Ty` -> `&'name mut Ty` or `&Ty` -> `&mut Ty`
+                    Some((
+                        mut_ty.ty.span.shrink_to_lo(),
+                        format!(
+                            "{}mut ",
+                            if lt.ident.span.lo() == lt.ident.span.hi() { "" } else { " " }
+                        ),
+                    ))
+                } else {
+                    None
+                }
+            })
+            .next()
+        else {
+            return false;
```

Comment:
can you not use like 10 different `let ... else ...` in this function and use a series of let chains instead like we do in `annotate_mut_binding_to_immutable_binding`. There's so much noise reading this with all the `else { return false; }` taking up multiple lines after practically every check in this function

Reply:
Not convinced that these long chains of conditions look better as let chains or let elses. I flip flop on my opinion by the hour 😅

Reply:
Yeah both always feel kind of gruelling to read lol...